mailet/standard/src/test/java/org/apache/james/transport/matchers/HasExceptionTest.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.mailet.base.test.FakeMatcherConfig;
+import org.apache.mailet.base.test.MailUtil;
+import org.junit.Before;
+import org.junit.Test;
+
+public class HasExceptionTest {
+
+    private FakeMail mockedMail;
+    private Matcher matcher;
+
+    @Before
+    public void setUp() throws Exception {
+        MimeMessage mimeMessage = MailUtil.createMimeMessage();
+        mockedMail = MailUtil.createMockMail2Recipients(mimeMessage);
+        matcher = new HasException();
+    }
+
+    @Test
+    public void matchShouldReturnAddressesWhenSpecifiedExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new javax.mail.internet.AddressException());
+
+        FakeMatcherConfig mci = FakeMatcherConfig.builder()
+                .matcherName("HasException")
+                .condition("javax.mail.internet.AddressException")
+                .build();
+
+        matcher.init(mci);
+
+        assertThat(matcher.match(mockedMail)).containsAll(mockedMail.getRecipients());
+    }
+
+    @Test
+    public void matchShouldReturnAddressesWhenSubclassOfSpecifiedExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new javax.mail.internet.AddressException());
+
+        FakeMatcherConfig mci = FakeMatcherConfig.builder()
+                .matcherName("HasException")
+                .condition("javax.mail.MessagingException")
+                .build();
+
+        matcher.init(mci);
+
+        assertThat(matcher.match(mockedMail)).containsAll(mockedMail.getRecipients());
+    }
+
+    @Test
+    public void matchShouldReturnNullWhenOtherExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new java.lang.RuntimeException());
+        
+        matcher.init(FakeMatcherConfig.builder()
+                .matcherName("HasException")
+                .condition("javax.mail.MessagingException")
+                .build());
+
+        assertThat(matcher.match(mockedMail)).isNull();
+    }
+
+    @Test
+    public void matchShouldReturnNullWhenSuperclassOfSpecifiedExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new javax.mail.MessagingException());
+
+        FakeMatcherConfig mci = FakeMatcherConfig.builder()
+                .matcherName("HasException")
+                .condition("javax.mail.internet.AddressException")
+                .build();
+
+        matcher.init(mci);
+
+        assertThat(matcher.match(mockedMail)).isNull();
+    }
+    
+    @Test
+    public void matchShouldReturnNullWhenNoExceptionHasOccurred() throws MessagingException {
+        matcher.init(FakeMatcherConfig.builder()
+                .matcherName("HasException")
+                .condition("java.lang.Exception")
+                .build());
+
+        assertThat(matcher.match(mockedMail)).isNull();
+    }
+
+}
```

Comment:
The test you wrote are really cool. I still think we can get it a bit further. I would propose to test:

 - What happens when the Mail.MAILET_ERROR_ATTRIBUTE_NAME is not an exception?

(Expected behaviour: not matching without error)

 - What happens when this matcher is configured with an invalid class name?
 - What happens when this matcher is configured with a class that is not an exception?